### PR TITLE
[Memhammer] Meta Schedule Rules

### PR DESCRIPTION
**Pull request by jinhongyii**
Migrated from tlc-pack repo

### CODE REVIEW

File: src/tir/schedule/primitive/read_write_at.cc [hunk 1]
```diff
@@ -134,7 +134,7 @@ struct ReadWriteAtImpl {
         impl.MakeLoopAndBlock<is_read>(src->name + "_" + storage_scope);
     StmtSRef result_block_sref =
         impl.ReplaceScopeBlock(new_loop_block.first.get(), new_loop_block.second->block.get());
-    impl.UpdateBlockInfo(result_block_sref);
+    impl.UpdateBlockInfo(result_block_sref, !new_loop_block.second->iter_values.empty());
```
Comment (vinx13) [inline src/tir/schedule/primitive/read_write_at.cc:137]: why is `affine_binding == !iter_values.empty()`?

File: src/arith/iter_affine_map.cc [hunk 2]
```diff
@@ -617,10 +615,17 @@ class IterMapRewriter : public ExprMutator {
     PrimExpr expected_extra_base = 0;
     PrimExpr expected_scale = base_scale.value();
     for (size_t i = 0; i < expr->args.size();) {
+      if (is_one(expr->args[i]->source->extent)) {
```
Comment (vinx13) [inline src/arith/iter_affine_map.cc:618]: can we have some test cases for iter map?

File: src/meta_schedule/schedule_rule/analysis.cc [hunk 3]
```diff
@@ -0,0 +1,194 @@
...
+  }
+  if (const auto* loop = pipeline_body_seq->seq[2].as<ForNode>()) {
+    if (loop->annotations.count(tir::attr::software_pipeline_stage)) {
+      return true;
+    }
+  }
+  return false;
+}
+
+void FallbackRule(const For& loop, Array<Integer>* stage, Array<Integer>* order) {
```
Comment (vinx13) [inline src/meta_schedule/schedule_rule/analysis.cc:138]:
```suggestion
void FallbackSoftwarPipelineRule(const For& loop, Array<Integer>* stage, Array<Integer>* order) {
```

File: src/meta_schedule/schedule_rule/analysis.h [hunk 4]
```diff
@@ -0,0 +1,160 @@
...
+ */
+std::vector<int> GetReadBufferNDims(const StmtSRef& block_sref);
+
+Optional<LoopRV> TilingwithTensorIntrin(const tir::Schedule& sch, const tir::BlockRV& block_rv,
+                                        const String& intrin_name);
+
+/*!
+ * \brief whether the loop's body has the pattern: 2 cache read shared followed by a nested
+ * software pipeline
+ * \param the loop
```
Comment (vinx13) [inline src/meta_schedule/schedule_rule/analysis.h:44]:
```suggestion
 * \param loop the loop
```

File: src/meta_schedule/schedule_rule/multi_level_tiling_memhammer.cc [hunk 5]
```diff
@@ -0,0 +1,390 @@
...
+    State new_state = state;
+    new_state.sch = sch;
+    if (new_state.tensor_core_is_used) new_state = TensorCoreLoad(new_state);
+    // add software pipeline annotations
+    tir::For loop = new_state.sch->Get(loop_rv);
+    Array<Integer> stage;
+    Array<Integer> order;
+    if (tir::IsCacheReadSharedPattern(loop)) {
+      stage = {0, 0, 0, 0, 0, 1, 1};
+      order = {0, 3, 1, 4, 5, 2, 6};
```
Comment (vinx13) [inline src/meta_schedule/schedule_rule/multi_level_tiling_memhammer.cc:335]: I'd prefer doing it with at least some basic analysis

**Comment by vinx13** [comment]
@jinhongyii can you rebase and let's get this in

**Comment by junrushao** [comment]
I'm going to close this PR given it looks like it's been stale for a while. @jinhongyii @vinx13

**Pull request closed without merging**